### PR TITLE
Move Swift publishing workflow to main repo

### PR DESCRIPTION
**Pull request by cnixbtc**
## Swift Publishing Workflow Consolidation

This PR relocates the Swift publishing workflow from the [`breez/breez-sdk-swift`](https://github.com/breez/breez-sdk-swift/blob/main/.github/workflows/publish-swift-package.yaml) repository to the main repository. This change simplifies the publishing process by integrating it with the main repo's CI workflow.

### 📖 Overview

Previously, the Swift publishing workflow required separate triggers, unlike other platforms managed centrally in the main repo via the [`publish-all-platforms`](https://github.com/breez/breez-sdk/blob/main/.github/workflows/publish-all-platforms.yml) workflow. This PR enhances the Swift package publishing by:

- Streamlining the publishing process by integrating the Swift package into the main repo's CI workflow.
- Improving build runtime and reliability by reusing the already built binary SDK and language bindings from the main repo's workflows.

One interesting point: Because our Flutter and React Native packages rely on the Swift package at runtime, the PR includes logic to ensure that the Flutter and React Native packages are only published if the Swift package is successfully published too, preventing issues with runtime dependencies. If the Swift package is not set to be published, then the Flutter and React Native packages are published nonetheless allowing us to release updates to those packages that depend on an already published Swift package. Special thanks to @JssDWt for the solution adapted from [this PR](https://github.com/breez/breez-sdk/pull/620).

### ❓ Outstanding Questions

Before merging this PR, we need to address two key points. The first is straightforward, while the second requires some feedback.

1. **Setting CocoaPods Token**  
   We need to set `secrets.COCOAPODS_TRUNK_TOKEN` in the main repo to publish to the CocoaPods trunk. This token is currently only set in the `breez-sdk-swift` repo (I believe we're using @roeierez's token there). The simplest solution is to just move the token from the Swift repo to the main repo.

2. **Publishing `breez_sdkFFI.xcframework`**  
   The Swift package and CocoaPods require a download link to the `breez_sdkFFI.xcframework` binary artifact. Currently, we create a GitHub release in the `breez-sdk-swift` repo and attach the XCFramework as a binary artifact. Moving the workflow to the main repo presents a few challenges for maintaining this setup. I believe options 1 and 3 are the most straightforward, with option 2 being a viable alternative that may need some experimentation.

   - **Option 1**: Use a GitHub Personal Access Token to replicate the current setup. This would involve creating a release in the `breez-sdk-swift` repo from the CI workflow in the main repo. While simple, this requires using user-bound tokens with broad permissions, which isn't ideal.
   - **Option 2**: Use a GitHub app to provide short-lived tokens for the workflow. This option offers better security with non-user-bound tokens but requires more setup.
   - **Option 3**: Create the release in the main `breez-sdk-greenlight` repo instead of the `breez-sdk-swift` repo and attach the artifact there. This option is easy to set up and doesn't require special tokens, but it requires releases to be created on the main repo whenever we want to update or publish the Swift package. Not ideal, imo.
   - **Option 4**: Find an alternative hosting solution outside of GitHub for the artifact, accessible to our users.

### ✋ Feedback Request

I'd appreciate your thoughts on these options, particularly regarding the second point. Looking forward to your feedback. 🙏

### CODE REVIEW

File: .github/workflows/publish-swift.yml
Comment (JssDWt) [inline .github/workflows/publish-swift.yml:134]: It would be nice to upload an artifact in this workflow, like we do for other packages, like [flutter](https://github.com/breez/breez-sdk-greenlight/blob/main/.github/workflows/publish-flutter.yml#L86-L92), [react-native](https://github.com/breez/breez-sdk-greenlight/blob/main/.github/workflows/publish-react-native.yml#L56-L60), [android](https://github.com/breez/breez-sdk-greenlight/blob/main/.github/workflows/publish-android.yml#L56-L60), etc. That way we can use that in the breez-sdk-docs to build docs independent of published versions.

**Review by JssDWt — COMMENTED** [review]
Looking good to me (always hard to say that with CI, you never know if it works)! I'd say if this is tested it's good to go.

**Comment by JssDWt** [comment]
> 2\. **Publishing `breez_sdkFFI.xcframework`**
>     The Swift package and CocoaPods require a download link to the `breez_sdkFFI.xcframework` binary artifact. Currently, we create a GitHub release in the `breez-sdk-swift` repo and attach the XCFramework as a binary artifact. Moving the workflow to the main repo presents a few challenges for maintaining this setup. I believe options 1 and 3 are the most straightforward, with option 2 being a viable alternative that may need some experimentation.

We have a similar setup for go. There we pass a repo ssh key to provide commit access. https://github.com/breez/breez-sdk-greenlight/blob/main/.github/workflows/publish-golang.yml#L31. I think we can use the same construct here (so option 1).

**Comment by dangeross** [comment]
Also check out the https://github.com/breez/breez-sdk-liquid repo, it publishes from the main publishing flow

**Comment by cnixbtc** [comment]
Thanks for the feedback guys and the pointer to the other repos. 

Will have a look!

**Comment by cnixbtc** [comment]
Alright, I've added an action-archival step for the XCFramework (the language bindings are already archived [here](https://github.com/breez/breez-sdk-greenlight/blob/main/.github/workflows/build-language-bindings.yml#L108-L113)).

That means we're now just missing the publishing token analogous to the publishing for the [Swift publishing for breez-sdk-liquid](https://github.com/breez/breez-sdk-liquid/blob/main/.github/workflows/publish-swift.yml#L133).

@yaslama Is that ok with you? I know we've previously talked about trying not to use a token. Let me know what you think. 🙏

**Comment by yaslama** [comment]
> @yaslama Is that ok with you? I know we've previously talked about trying not to use a token. Let me know what you think. 🙏

Let's use tokens for now and try to find an alternative.

**Comment by JssDWt** [comment]
Looks good it me, if it works, let's merge it. Did you test this with `publish=false` with some different configurations?

File: .github/workflows/publish-swift.yml
Comment (dangeross) [inline .github/workflows/publish-swift.yml:92]: This is causing the publish CI to fail when testing
```suggestion
          name: breez_sdkFFI-${{ inputs.package-version || input.ref }}.xcframework
```

File: .github/workflows/publish-swift.yml
Comment (dangeross) [inline .github/workflows/publish-swift.yml:124]: I seem to remember this action had issues in the [breez-sdk-liquid](https://github.com/breez/breez-sdk-liquid) repo when using a token. There we use `softprops/action-gh-release@v2` as the release action instead.

File: .github/workflows/publish-swift.yml
Comment (cnixbtc) [reply .github/workflows/publish-swift.yml:92]: Damn, typo! Thanks for catching!

File: .github/workflows/publish-swift.yml
Comment (cnixbtc) [reply .github/workflows/publish-swift.yml:124]: Good insight thanks! Will use `softprops/action-gh-release@v2` here too. Good for consistency too! 👏

**Comment by cnixbtc** [comment]
> Looks good it me, if it works, let's merge it. Did you test this with `publish=false` with some different configurations?

Thanks for your feedback! 🙇 I primarily tested the syncing logic to ensure that the React Native and Flutter publishing workflows behave as follows:

- Both Swift and Android packages are published: React Native and Flutter are published only if both Swift and Android publishing succeed.
- Either Swift or Android package is published: React Native and Flutter are published only if the respective (Swift or Android) publishing succeeds.
- Neither Swift nor Android packages are published: React Native and Flutter are published regardless of Swift or Android status, allowing for RN/Flutter patches without re-publishing Swift or Android.

<details>
  <summary>Here's my test notes: </summary>
<ul>
    <li>Tests using dummy binaries (publish == false)
        <ul>
            <li>Wokflow config: ["maven", "swift"]
                <ul>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9928882340">https://github.com/breez/breez-sdk/actions/runs/9928882340</a></li>
                    <li>jobs dependent on maven and swift run after maven and swift as expected</li>
                </ul>
            </li>
            <li>Wokflow config: ["maven"]
                <ul>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9928828401">https://github.com/breez/breez-sdk/actions/runs/9928828401</a></li>
                    <li>jobs dependent on maven and swift run after maven as expected</li>
                </ul>
            </li>
            <li>Wokflow config: ["swift"]
                <ul>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9928935630">https://github.com/breez/breez-sdk/actions/runs/9928935630</a></li>
                    <li>jobs dependent on maven and swift run after swift as expected</li>
                </ul>
            </li>
            <li>Wokflow config: ["maven", "swift"]
                <ul>
                    <li>hardcoded a fail for swift</li>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9928994716">https://github.com/breez/breez-sdk/actions/runs/9928994716</a></li>
                    <li>jobs dependent on maven and swift don't run as expected</li>
                </ul>
            </li>
            <li>Wokflow config: ["golang"]
                <ul>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9929068463">https://github.com/breez/breez-sdk/actions/runs/9929068463</a></li>
                    <li>jobs dependent on maven and swift run after as expected</li>
                </ul>
            </li>
        </ul>
    </li>
    <li>Tests using actual binaries (publish == false)
        <ul>
            <li>Wokflow config: ["flutter", "swift"]
                <ul>
                    <li><a href="https://github.com/breez/breez-sdk/actions/runs/9929607744">https://github.com/breez/breez-sdk/actions/runs/9929607744</a></li>
                    <li>jobs dependent on maven and swift run after maven and swift as expected</li>
                </ul>
            </li>
        </ul>
    </li>
</ul>
</details>

Of course, CI is hard to fully test so when we merge this and see that publishing fails feel free to revert and let me know and I will fix whatever problems there might be.

**Comment by cnixbtc** [comment]
⚠️ Before we merge this, we need to set the `SWIFT_RELEASE_TOKEN` for the https://github.com/breez/breez-sdk-swift repo as a GH actions secret here.

**Comment by JssDWt** [comment]
> Here's my test notes:
> * Tests using dummy binaries (publish == false)
>           
>               * Wokflow config: ["maven", "swift"]
>                     
>                         * https://github.com/breez/breez-sdk/actions/runs/9928882340
>                         * jobs dependent on maven and swift run after maven and swift as expected
>               * Wokflow config: ["maven"]
>                     
>                         * https://github.com/breez/breez-sdk/actions/runs/9928828401
>                         * jobs dependent on maven and swift run after maven as expected
>               * Wokflow config: ["swift"]
>                     
>                         * https://github.com/breez/breez-sdk/actions/runs/9928935630
>                         * jobs dependent on maven and swift run after swift as expected
>               * Wokflow config: ["maven", "swift"]
>                     
>                         * hardcoded a fail for swift
>                         * https://github.com/breez/breez-sdk/actions/runs/9928994716
>                         * jobs dependent on maven and swift don't run as expected
>               * Wokflow config: ["golang"]
>                     
>                         * https://github.com/breez/breez-sdk/actions/runs/9929068463
>                         * jobs dependent on maven and swift run after as expected
> * Tests using actual binaries (publish == false)
>           
>               * Wokflow config: ["flutter", "swift"]
>                     
>                         * https://github.com/breez/breez-sdk/actions/runs/9929607744
>                         * jobs dependent on maven and swift run after maven and swift as expected

Great notes!

**Comment by JssDWt** [comment]
> ⚠️ Before we merge this, we need to set the `SWIFT_RELEASE_TOKEN` for the https://github.com/breez/breez-sdk-swift repo as a GH actions secret here.
> 
> Who can help me with that?

@roeierez

File: .github/workflows/publish-swift.yml
Comment (JssDWt) [inline .github/workflows/publish-swift.yml:92]: Still a typo!

```suggestion
          name: breez_sdkFFI-${{ inputs.package-version || inputs.ref }}.xcframework
```

Comment (cnixbtc) [reply .github/workflows/publish-swift.yml]: Haha damn. Sorry about that.

**Comment by roeierez** [comment]
> > ⚠️ Before we merge this, we need to set the `SWIFT_RELEASE_TOKEN` for the https://github.com/breez/breez-sdk-swift repo as a GH actions secret here.
> > Who can help me with that?
> 
> @roeierez

@JssDWt @cnixbtc Done.

**Review by JssDWt — APPROVED** [review]
LGTM!

File: .github/workflows/publish-swift.yml
Comment (dangeross) [inline .github/workflows/publish-swift.yml:26]:
```suggestion
      SWIFT_RELEASE_TOKEN:
```

File: .github/workflows/publish-all-platforms.yml
Comment (dangeross) [inline .github/workflows/publish-all-platforms.yml:319]:
```suggestion
      SWIFT_RELEASE_TOKEN: ${{ secrets. SWIFT_RELEASE_TOKEN }}
      COCOAPODS_TRUNK_TOKEN: ${{ secrets.COCOAPODS_TRUNK_TOKEN }}
```

**Review by dangeross — CHANGES_REQUESTED** [review]

**Comment by cnixbtc** [comment]
Thanks for the corrections @dangeross. Note to myself: No more last minute changes without proofreading. Will be more careful with these things next time.

Just to double check: Is the `COCOAPODS_TRUNK_TOKEN` already set as a secret in this repo? If not I think @roeierez will need to add that one too.

**Comment by cnixbtc** [comment]
> Thanks for the corrections @dangeross. Note to myself: No more last minute changes without proofreading. Will be more careful with these things next time. I'll give it one more trial run here: https://github.com/breez/breez-sdk-greenlight/actions/runs/10706106599
> 
> Just to double check: Is the `COCOAPODS_TRUNK_TOKEN` already set as a secret in this repo? If not I think @roeierez will need to add that one too.

Test run failed due to some (I think) unrelated issues with building the Kotlin lang bindings. But I saw that @JssDWt did some test runs earlier that were successful. I will be on holiday for the next 2 weeks (starting on the 7th). Let me know guys if you'd prefer to merge this before or after. :)

**Comment by JssDWt** [comment]
> Test run failed due to some (I think) unrelated issues with building the Kotlin lang bindings. But I saw that @JssDWt did some test runs earlier that were successful. I will be on holiday for the next 2 weeks (starting on the 7th). Let me know guys if you'd prefer to merge this before or after. :)

This is a problem that comes up after rust 1.80. I ran into this issue with cln as well, I downgraded to rust 1.79 there. Let me see how to fix this here.

**Comment by JssDWt** [comment]
@cnixbtc can you try a rebase?

**Comment by cnixbtc** [comment]
> @cnixbtc can you try a rebase?

Rebased onto main.

**Comment by cnixbtc** [comment]
Guys, as mentioned above, I will be off the grid for 2 weeks now. I can take this over the finish line and when I'm back but of course feel free to merge it beforehand when you feel like it. Whatever it is thanks for the help so far and I'll report back in two weeks. :)

**Comment by JssDWt** [comment]
I've removed the rust caches, and now the build works. It was picking `time` 0.3.30`, where it should have picked `0.3.36`. Or maybe the build didn't use your rebased version yet.

**Comment by cnixbtc** [comment]
Alright, so I did three more test runs. For all three I used the workflow from this branch and the repo code from the 0.6.1 tag.

1. [`["flutter", "react-native", "swift"]`](https://github.com/breez/breez-sdk-greenlight/actions/runs/11070496392): swift runs first, then flutter and react native in parallel
2. [`["react-native", "flutter"]`](https://github.com/breez/breez-sdk-greenlight/actions/runs/11070708899): both run in parallel without waiting for the (non existing) swift build
3. [`["swift"]`](https://github.com/breez/breez-sdk-greenlight/actions/runs/11071035095): runs as expected

Good to go from my side.

**Review by JssDWt — APPROVED** [review]
LGTM

**Comment by dangeross** [comment]
> Just to double check: Is the `COCOAPODS_TRUNK_TOKEN` already set as a secret in this repo? If not I think @roeierez will need to add that one too.

@roeierez

**Comment by cnixbtc** [comment]
@dangeross I checked with Roei, `COCOAPODS_TRUNK_TOKEN` is set.

`SWIFT_RELEASE_TOKEN` is set too and `REPO_SSH_KEY` is used by other jobs too so must be set already. 🙃

**Review by dangeross — APPROVED** [review]
LGTM

**Pull request merged**
